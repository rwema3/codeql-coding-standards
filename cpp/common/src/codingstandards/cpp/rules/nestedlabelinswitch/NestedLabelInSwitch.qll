/**
 * Provides a library which includes a `problems` predicate for reporting nested labels in a switch statement.
 */

import cpp
import codingstandards.cpp.Customizations
import codingstandards.cpp.Exclusions

abstract class NestedLabelInSwitchSharedQuery extends Query { }

Query getQuery() { result instanceof NestedLabelInSwitchSharedQuery }

query predicate problems(
  SwitchCase case, string message, SwitchCase caseLocation, string caseLabel, SwitchStmt switch,
  string switchLabel
) {
  not isExcluded(case, getQuery()) and
  switch.getASwitchCase() = caseLocation and
  not case.getParentStmt() = switch.getChildStmt() and
  case = caseLocation and
  message =
    "The case $@ does not appear at the outermost level of the compound statement forming the body of the $@ statement." and
  caseLabel = case.toString() and
  switchLabel = switch.toString()
}
