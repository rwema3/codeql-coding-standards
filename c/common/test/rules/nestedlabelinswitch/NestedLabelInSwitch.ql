// GENERATED FILE - DO NOT MODIFY
import codingstandards.cpp.rules.nestedlabelinswitch.NestedLabelInSwitch

class TestFileQuery extends NestedLabelInSwitchSharedQuery, TestQuery { }
