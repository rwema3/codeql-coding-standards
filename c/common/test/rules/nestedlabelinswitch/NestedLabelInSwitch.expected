| test.c:9:5:9:11 | case ...: | The case $@ does not appear at the outermost level of the compound statement forming the body of the $@ statement. | test.c:9:5:9:11 | case ...: | case ...: | test.c:6:3:17:3 | switch (...) ...  | switch (...) ...  |
| test.c:36:5:36:11 | case ...: | The case $@ does not appear at the outermost level of the compound statement forming the body of the $@ statement. | test.c:36:5:36:11 | case ...: | case ...: | test.c:23:3:43:3 | switch (...) ...  | switch (...) ...  |
| test.c:76:5:76:11 | case ...: | The case $@ does not appear at the outermost level of the compound statement forming the body of the $@ statement. | test.c:76:5:76:11 | case ...: | case ...: | test.c:73:3:79:3 | switch (...) ...  | switch (...) ...  |
